ironfish/src/network/peerNetwork.ts

```diff
@@ -718,11 +720,73 @@ export class PeerNetwork {
     return new PooledTransactionsResponse(transactions, rpcId)
   }
 
-  private onGetBlockTransactionsRequest(
+  private async onGetBlockTransactionsRequest(
+    peer: Peer,
     message: GetBlockTransactionsRequest,
-  ): GetBlockTransactionsResponse {
-    // TODO(IRO-2279): Implement a handler for this
-    return new GetBlockTransactionsResponse(message.blockHash, [], message.rpcId)
+  ): Promise<GetBlockTransactionsResponse> {
+    const block = await this.chain.db.withTransaction(null, async (tx) => {
+      const header = await this.chain.getHeader(message.blockHash, tx)
+
+      if (header === null) {
+        throw new CannotSatisfyRequestError(
+          `Peer requested transactions for block ${message.blockHash.toString(
+            'hex',
+          )} that isn't in the database`,
+        )
+      }
+
+      if (header.sequence < this.chain.head.sequence - MAX_GET_BLOCK_TRANSACTIONS_DEPTH) {
+        throw new CannotSatisfyRequestError(
+          `Peer requested transactions for block ${message.blockHash.toString(
+            'hex',
+          )} with sequence ${header.sequence} while chain head is at sequence ${
+            this.chain.head.sequence
+          }`,
+        )
+      }
+
+      const block = await this.chain.getBlock(header, tx)
+
+      Assert.isNotNull(
+        block,
+        'Database should contain transactions if it contains block header',
+      )
+
+      return block
+    })
+
+    if (message.transactionIndexes.length > block.transactions.length) {
+      const errorMessage = `Requested ${
+        message.transactionIndexes.length
+      } transactions for block ${block.header.hash.toString('hex')} that contains ${
+        block.transactions.length
+      } transactions`
+      peer.punish(BAN_SCORE.MAX, errorMessage)
+      throw new CannotSatisfyRequestError(errorMessage)
+    }
+
+    const transactions = []
+    let currentIndex = 0
+    for (const transactionIndex of message.transactionIndexes) {
+      if (transactionIndex < 0) {
+        const errorMessage = `Requested negative transaction index`
+        peer.punish(BAN_SCORE.MAX, errorMessage)
+        throw new CannotSatisfyRequestError(errorMessage)
+      }
```

Comment:
This shouldn't practically ever happen, since bufio's `readVarint` only allows for unsigned integers, but it's not easy to propagate that assumption through the code.